autoextract_poet/providers.py

```diff
@@ -0,0 +1,49 @@
+from typing import Type
+
+from scrapy.http import Request
+from scrapy.statscollectors import StatsCollector
+from scrapy_poet.page_input_providers import (
+    PageObjectInputProvider,
+    provides,
+)
+
+from autoextract_poet.inputs import (
+    ProductResponseData,
+    ProductListResponseData,
+)
+from autoextract_poet.utils import request
+
+
+class AutoExtractProvider(PageObjectInputProvider):
+
+    page_type: str
+    page_input_class: Type
+
+    def __init__(self, request: Request, stats: StatsCollector):
+        self.request = request
+        self.stats = stats
+
+    async def __call__(self):
+        try:
+            # FIXME: how to configure if you want full HTML or not?
+            data = await request(self.request.url, self.page_type)
+        except Exception:
+            self.stats.inc_value(f"autoextract/{self.page_type}/error")
+            raise
+
+        self.stats.inc_value(f"autoextract/{self.page_type}/success")
+        return self.page_input_class(data=data)
+
+
+@provides(ProductResponseData)
+class AutoExtractProductResponseDataProvider(AutoExtractProvider):
+
+    page_type = "product"
+    page_input_class = ProductResponseData
+
+
+@provides(ProductListResponseData)
+class AutoExtractProductListResponseDataProvider(AutoExtractProvider):
+
+    page_type = "productList"
+    page_input_class = ProductListResponseData
```

Comment:
`ProductListResponseData` has to be repeated twice here. What if we allow injecting the registered type so that it is received in the init. I mean creating an object:
```py
@dataclass
class ClassToProvide:
  cls: Type
```

And modify the Middleware so that you can receive it in the `__init__`:
```py
class AutoExtractProvider(PageObjectInputProvider):
    page_type: str

    def __init__(self, request: Request, stats: StatsCollector, cls_to_provide: ClassToProvide):
        self.request = request
        self.stats = stats
        self.page_input_class = cls_to_provide.cls

   ...
```

Reply:
I guess it's a good idea, just don't know if "too magic", like hard to trace where things are happening.
Also, what would happen if the user requests for a class that is not available in AutoExtract (something that is not a product, listing, ...)?

Reply:
There's also some indication from @kmike that we should be registering those providers manually (probably in the Scrapy Settings).